### ISSUE TITLE
std_hw / var_hw / normalize_hw: the variance divides by the padded element count and by N, where the registered torch.std golden divides by logical N-1

### ISSUE THREAD

**Issue by botify-service** [tenstorrent/tt-metal#53277]
### Component / Area

ttnn eltwise unary composites — `ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp` (`_variance_impl`, and `var_hw` / `std_hw` / `normalize_hw` / `normalize_global` built on it)

### Issue Type (optional)

Bad Outputs

### Observed

`_variance_impl` divides the sum of squared deviations by the wrong element count. Two separate
defects, both in the one `scale` expression, and both visible in the source without running
anything:

```cpp
// unary_composite_op.cpp, _variance_impl
constexpr float correction = 0.0f;
auto shape_wh = y.padded_shape();
float scale = 1.0f / ((float)(shape_wh[3] * shape_wh[2]) - correction);
```

**1. The divisor uses `padded_shape()` while the mean it subtracts uses the logical shape.**

`var_hw` computes the mean one line earlier with `ttnn::mean(y, {2,3}, true)`. That path divides by
the *logical* extent — `generic_reductions.cpp` takes `input_tensor_arg.logical_shape()` and
accumulates `reduced_volume *= input_shape[axis]`. `_variance_impl` then divides the squared
deviations by the *padded* extent. The two halves of the same variance disagree about how many
elements there are, so for any tensor whose H or W is not a multiple of 32 the variance is scaled
by `logical_HW / padded_HW`.

**2. `correction` is pinned to `0.0f`, but the registered golden is `torch.std`.**

`_golden_function_normalize_hw` in `ttnn/ttnn/operations/unary.py` is:

```python
mean_hw = torch.mean(input_tensor_a, [-2, -1], keepdim=True)
std_hw  = torch.std(input_tensor_a, [-2, -1], keepdim=True)
```

`torch.std` defaults to `correction=1` — the sample standard deviation, dividing by `N-1`. The
composite computes the population value, dividing by `N`. This one is unconditional: it applies to
every call, including perfectly tile-aligned ones where defect 1 is inert.

The modern reduction path already models this correctly — `generic_reductions.cpp` carries a real
`correction` parameter and computes `divisor = correction ? (reduced_volume - 1) : reduced_volume`
off the logical shape. The `_hw` / `_global` composites never got that treatment.

### Magnitude

`std_hw` returns `sqrt(padded_HW / (logical_HW - 1))` times the value its golden returns.

Defect 2 alone, on tile-aligned shapes where padding is irrelevant:

| logical H×W | N | ttnn divides by | `torch.std` divides by | ratio | error |
|---|---|---|---|---|---|
| 32×32 | 1024 | 1024 | 1023 | 1.000489 | 0.049% |
| 32×64 | 2048 | 2048 | 2047 | 1.000244 | 0.024% |
| 64×64 | 4096 | 4096 | 4095 | 1.000122 | 0.012% |
| 256×256 | 65536 | 65536 | 65535 | 1.000008 | 0.001% |

Small, but one-directional — it never averages out, and it is a divergence from the registered
reference rather than a rounding question.

Both defects together, on logical shapes that are not tile multiples:

| logical H×W | N | padded N | ratio | error |
|---|---|---|---|---|
| 32×1 | 32 | 1024 | 5.747 | 475% |
| 7×7 | 49 | 1024 | 4.619 | 362% |
| 5×17 | 85 | 1024 | 3.492 | 249% |
| 16×16 | 256 | 1024 | 2.004 | 100% |
| 33×33 | 1089 | 4096 | 1.940 | 94% |
| 100×100 | 10000 | 16384 | 1.280 | 28% |
| 31×31 | 961 | 1024 | 1.033 | 3.3% |

`normalize_hw` divides by that standard deviation, so its output is wrong by the reciprocal — for a
logical `32×1`, every element comes back about `0.174×` what it should be.

There is a second-order effect in the same direction that I have *not* quantified: `y_minus_mean_y`
is formed by broadcasting over the full tile, so the padding lanes hold `0 - mean = -mean` and
contribute `mean²` each to the sum. For a zero-mean input that term vanishes and the table above is
the whole story; for a non-zero mean it adds on top. I am leaving it out of the numbers rather than
asserting a size for it.

### Expected

`std_hw` should match its registered golden `torch.std(x, [-2,-1])` to the op's tolerance, which
means dividing by `logical_HW - 1`, not by `padded_HW`. Equivalently, the two invariants that pin
this down without reference to any golden:

- `std_hw` of a tensor should not change when the tensor is padded to the next tile boundary. Today
  it does — that is defect 1 stated as an oracle.
- `var_hw(x) * (N-1) / N` should equal the population variance computed directly. Today `var_hw`
  *is* the population variance, off by that factor from what the golden returns.

`var_hw`, `normalize_hw` and `normalize_global` share `_variance_impl` and inherit both.

If the population variance is the intended semantics, then the divergence is in the golden rather
than the kernel, and the fix is to register `torch.std(..., correction=0)` and document it — but the
`padded_shape()` half is wrong either way, since no reading of "variance" divides by padding.

### Steps to reproduce

I do not have a Tenstorrent card. These are the commands for someone who does.

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

# Defect 2 — tile-aligned, so padding plays no part.
x = torch.randn((1, 1, 32, 32), dtype=torch.float32)
tx = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
got = ttnn.to_torch(ttnn.from_device(ttnn.std_hw(tx)))[0, 0, 0, 0].item()
want = torch.std(x, [-2, -1]).item()            # correction=1, the registered golden
print(f"32x32   ttnn={got:.8f}  golden={want:.8f}  ratio={got/want:.6f}")  # expect ~0.99951

# Defect 1 — logical shape is not a tile multiple.
y = torch.randn((1, 1, 7, 7), dtype=torch.float32)
ty = ttnn.from_torch(y, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
got = ttnn.to_torch(ttnn.from_device(ttnn.std_hw(ty)))[0, 0, 0, 0].item()
want = torch.std(y, [-2, -1]).item()
print(f"7x7     ttnn={got:.8f}  golden={want:.8f}  ratio={got/want:.6f}")  # expect ~0.2165

ttnn.close_device(device)
```

A correct implementation prints `ratio` of `1.0` on both lines.

### Why the current tests do not catch it

- The `std_hw` sweep compares with `comp_allclose(atol=0.1)`, raised by PRs #7708 / #7748. A 0.049%
  relative divergence on inputs drawn from `gen_rand(low=-10, high=10)` sits far inside that.
- Every sweep config for these ops — `ttnn_std_hw_test.yaml`, `ttnn_normalize_hw_test.yaml`,
  `ttnn_normalize_global_test.yaml` — uses tile-aligned shapes only, so defect 1 is never
  exercised. `start-shape: [1, 1, 32, 32]` stepping by `[1, 1, 32, 32]` cannot produce a logical
  extent that is not a multiple of 32.
- `tests/ttnn/docs_examples/test_eltwise_unary_examples.py::test_var_hw` is the only direct test of
  `var_hw` and asserts nothing.
- `test_reduction.py::test_std_var_hw_reduce_batch_crosses_tree_block` exercises `ttnn.std` /
  `ttnn.var` — the reduction path, which is correct — not these composites.

Neither an unpadded shape nor an exact-value comparison against the golden appears anywhere, which
is the gap that leaves both defects invisible.

### Relationship to #52614

Not a duplicate, and worth being explicit about the boundary. #52614 (open) covers the *range* of
the squared intermediate in this same `_variance_impl` — flush-to-zero below `sqrt(FLT_MIN)` and
saturation above `sqrt(2^127)`. This issue is about the *divisor*: which element count the sum is
divided by. The two are orthogonal — a fix that factors the magnitude out before squaring, as
#52614 describes, corrects the range and leaves `padded_shape()` and `correction = 0.0f` exactly as
they are. They do want fixing in the same function, and whoever takes one should probably see the
other.

### Is this a regression?

No. `_variance_impl` has computed the scale this way at the SHA below; I did not bisect and have no
evidence of a commit that changed it. Context, not blame: #25100 / PR #41838 (merged 2026-04-14)
reimplemented `ttnn.std` / `ttnn.var` with Welford under `operations/reduction/generic`, which is
where the correct `correction` and `logical_shape` handling lives today. That work did not touch
`unary_composite_op.cpp`, so the `_hw` and `_global` composites kept the older scale expression.

### Software Versions

| | |
|---|---|
| tt-metal | `main`, source as of 2026-08-15 |
| reference | `torch.std` semantics per the registered golden in `ttnn/ttnn/operations/unary.py` |

### Hardware Details

No Tenstorrent card, and no ttsim run — saying so up front rather than letting a reviewer find it.

Everything above is derived from the source: the `scale` expression in `_variance_impl`, the
logical-shape accumulation in `generic_reductions.cpp`, and the golden in `unary.py`. The ratio
tables are exact arithmetic on those element counts, not measurements — `sqrt(padded_HW/(N-1))`
evaluated per shape. The claim that `correction = 0.0f` diverges from `torch.std` needs no hardware,
since both sides are fixed by their definitions. The claim about unpadded shapes assumes a logical
extent below the tile boundary reaches `_variance_impl` with `padded_shape()` rounded up to 32,
which is the documented tiling behaviour but which I have not observed executing; a run of the
second block in the repro would settle it, and I would rather have that confirmation than assert it.

### Impact

`normalize_hw` and `normalize_global` sit in feature preprocessing and normalisation paths. The
tile-aligned error is small enough to pass every existing tolerance while biasing every normalised
tensor in one direction. The unpadded-shape error is not small, and it is silent — the output stays
finite and plausibly shaped, so nothing downstream flags it.

**Comment by botify-service** [comment]
@jberkowitzTT — is this in scope for the bounty program? I'd rather ask than assume a
price by titling it myself.

Scoping notes in case they help the review:

- The `correction = 0.0f` half is a one-line divergence from the registered golden and
  affects every call. The `padded_shape()` half is the larger error but only bites on
  logical shapes that are not tile multiples.
- It is orthogonal to #52614, which covers the range of the squared intermediate in the
  same `_variance_impl`. A fix for that one leaves this divisor untouched, so the two
  are separable — though whoever takes either should probably see the other.
- I have no Tenstorrent card, so the tables are exact arithmetic on element counts
  rather than measurements. That is stated in the issue, and the repro is written for
  someone with hardware.

Happy to take it if it is ratified, and equally happy for it to be folded into #52614
if you would rather have one fix for that function.
